gneissweb_classification

ran-iwamoto opened the pull request:
## Why are these changes needed?
This PR adds a transform for gneissweb classification using fasttext classifier.

## Related issue number (if any).
issue #924

ran-iwamoto committed:
gneissweb_classification

ran-iwamoto commented:
Please check @shahrokhDaijavad @touma-I

shahrokhDaijavad committed:
Made a few changes in README

Signed-off-by: SHAHROKH DAIJAVAD <shahrokh@us.ibm.com>

shahrokhDaijavad commented:
@touma-I and @ran-iwamoto I reviewed and modified the README file. It looks good now. I also tested both Python and Ray versions of notebook (after creating my own HF token and using that in the notebooks) successfully. CI/CD is failing above, because test-language-gneissweb_classification.yml is missing.

Swanand-Kadhe commented:
One point for discussion: I was wondering if we want to call this transform as `gneissweb classification` transform. This is a more generic transform for `fastText classification`. We will use this transform with the fastText models used in GneissWeb to create GneissWeb recipe notebook. However, one can use this transform outside of GneissWeb. For instance, one can reproduce fastText classification step of DCLM dataset using this transform with the appropriate fastText model. It would be good to know what others think. Many thanks!

shahrokhDaijavad commented:
> One point for discussion: I was wondering if we want to call this transform as `gneissweb classification` transform. This is a more generic transform for `fastText classification`. We will use this transform with the fastText models used in GneissWeb to create GneissWeb recipe notebook. However, one can use this transform outside of GneissWeb. For instance, one can reproduce fastText classification step of DCLM dataset using this transform with the appropriate fastText model. It would be good to know what others think. Many thanks!

Thanks, @Swanand-Kadhe Your point is well-taken. However, we have discussed quite a bit about the naming of this transform in #924 and the name `gneissweb_classification` has been chosen, until we come up with a better name. The name `fastText classification` is too broad for what we have now.,

touma-I committed:
generate workflow script

Signed-off-by: Maroun Touma <touma@us.ibm.com>

shahrokhDaijavad commented on transforms/language/gneissweb_classification/README.md:
```diff
@@ -0,0 +1,78 @@
+| _content_column_name_ | `contents` | specifies name of the column containing documents |
+| _output_lablel_column_name_ | `label` | specifies name of the output column to hold predicted classes|
+| _output_score_column_name_ | `score` | specifies name of the output column to hold score of prediction |
+
```
@ran-iwamoto A user will wonder what the `__gcls` is doing in front of all the options. Doesn't it make sense to use the same name for the parameter and the command line option for each key? If you make the change here, please make sure, you make the corresponding change in the the 2 notebooks.

ran-iwamoto replied:
Thank you for review, I understand your point. But the current structure is aligned with the design of the language identification transform. Wouldn't it make sense to maintain consistency with that structure instead of deviating from it?
I think the problem is hard to understand short name?

touma-I replied:
@ran-iwamoto This document  is meant for the end users and it should describe the configuration parameters (gcls_*) that are used when invoking the Classification API

shahrokhDaijavad replied:
OK, thanks for the explanation, @ran-iwamoto. So, `lang_id_` has been replaced by `gcls_` in the option name. What does gcls stand for? gneissweb classification? Maybe adding a short sentence explaining that we add gcls prefix to each parameter, which stands for gcls classification, would be helpful.

ran-iwamoto replied:
I added the sentence in read me "The prefix gcls is short name for Gneissweb CLaSsification.".

shahrokhDaijavad reviewed (changes requested):
See inline comment.

touma-I commented on transforms/language/gneissweb_classification/test/test_gneissweb_classification.py:
This UT is failing. Please see: https://github.com/IBM/data-prep-kit/actions/runs/12998928593/job/36253146761?pr=974
```
==================================== ERRORS ====================================
_ ERROR collecting language/gneissweb_classification/test/test_gneissweb_classification.py _
test_gneissweb_classification.py:131: in get_test_transform_fixtures
    ClassificationTransform(config),
../dpk_gneissweb_classification/transform.py:64: in __init__
    self.nlp_classfication = self._get_nlp_classfication(config)
../dpk_gneissweb_classification/transform.py:80: in _get_nlp_classfication
    raise ValueError("model_file_name_key is not specified.")
E   ValueError: model_file_name_key is not specified.
=========================== short test summary info ============================
ERROR test_gneissweb_classification.py::TestLangIdentificationTransform - ValueError: model_file_name_key is not specified.
```

ran-iwamoto replied:
I changed parameters of two files from filename to file_name. Test will success.

ran-iwamoto committed:
Update README.md

The sentence is added "The prefix gcls is short name for Gneissweb CLaSsification.".

ran-iwamoto committed:
Update test_gneissweb_classification.py

ran-iwamoto committed:
Update test_gneissweb_classification_python.py

Swanand-Kadhe commented:
> Thanks, @Swanand-Kadhe Your point is well-taken. However, we have discussed quite a bit about the naming of this transform in #924 and the name `gneissweb_classification` has been chosen, until we come up with a better name. The name `fastText classification` is too broad for what we have now.,

Thanks a lot, @shahrokhDaijavad, for pointing me to the discussion in #924. I should have looked at it earlier. Sorry about that.

shahrokhDaijavad commented:
Thank you, @ran-iwamoto, for making the latest changes. Let's wait for the other reviewers also to look before we approve and merge.

touma-I commented on transforms/language/gneissweb_classification/README.md:
replace "key name" with configuration parameters gcls_*.

cc: @shahrokhDaijavad

touma-I commented on transforms/language/gneissweb_classification/gneissweb_classification.ipynb:
Replace "key name" with configuration parameters gcls_*

cc: @shahrokhDaijavad

touma-I commented on transforms/language/gneissweb_classification/gneissweb_classification-ray.ipynb:
Replace "key name" with configuration parameters gcls_*

cc: @shahrokhDaijavad

ran-iwamoto committed:
Update README.md

ran-iwamoto committed:
Update gneissweb_classification-ray.ipynb

cmadam commented on transforms/language/gneissweb_classification/dpk_gneissweb_classification/transform.py:
```diff
@@ -0,0 +1,171 @@
+    def _get_nlp_classfication(config) -> ClassificationModel:
+        nlp_classfication: ClassificationModel
+
+        model_credential = config.get(model_credential_key)
```
I would recommend to use CLI parameters in both the transform and the transform configuration. To achieve this behavior, please use `*_cli_param` instead of `*_key`.  For example, instead of:
```python
        model_credential = config.get(model_credential_key)
        model_file_name = config.get(model_file_name_key)
        model_url = config.get(model_url_key)
```
you can use
```python
        model_credential = config.get(model_credential_cli_param)
        model_file_name = config.get(model_file_name_cli_param)
        model_url = config.get(model_url_cli_param)
```

cmadam commented on transforms/language/gneissweb_classification/dpk_gneissweb_classification/transform.py:
```diff
@@ -0,0 +1,171 @@
+        :return: True, if validate pass or False otherwise
+        """
+        captured = CLIArgumentProvider.capture_parameters(args, cli_prefix, False)
+        self.params = self.params | captured
```
To be able to pass `*_cli_param` instead of `*_key`, you should also replace:
```python
        captured = CLIArgumentProvider.capture_parameters(args, cli_prefix, False)
```
with
```python
        captured = CLIArgumentProvider.capture_parameters(args, cli_prefix, True)
```

cmadam reviewed (dismissed):
Great implementation, however, I would suggest to use `*_cli_param` instead of `*_key` configuration keys in transform, as highlighted in the two comments below.

ran-iwamoto committed:
Update transform.py key to cli_param

ran-iwamoto committed:
Update transform.py To be able to pass *_cli_param instead of *_key, should replace

shahrokhDaijavad committed:
A few changes in README and notebooks, related to config parameters

Signed-off-by: SHAHROKH DAIJAVAD <shahrokh@us.ibm.com>

shahrokhDaijavad commented:
@ran-iwamoto I pushed a few changes to your branch, to finalize the parameter names and configuration in README and the comment parts of the two notebooks.

shahrokhDaijavad commented:
@BishwaBhatta This transform has been tested only with `facebook/fasttext-language-identification model.bin `
Whenever we have the models that were used in creating GneissWeb fasttext classification uploaded to HF or a similar public place, we should test it again.

touma-I commented on transforms/language/gneissweb_classification/test/test_gneissweb_classification.py:
This UT is failing. Please fix it. Thanks
```
==================================== ERRORS ====================================
_ ERROR collecting language/gneissweb_classification/test/test_gneissweb_classification.py _
test_gneissweb_classification.py:131: in get_test_transform_fixtures
    ClassificationTransform(config),
../dpk_gneissweb_classification/transform.py:64: in __init__
    self.nlp_classfication = self._get_nlp_classfication(config)
../dpk_gneissweb_classification/transform.py:78: in _get_nlp_classfication
    raise ValueError("model_credential_cli_param is not specified.")
E   ValueError: model_credential_cli_param is not specified.
=========================== short test summary info ============================
ERROR test_gneissweb_classification.py::TestLangIdentificationTransform - ValueError: model_credential_cli_param is not specified.
```

ran-iwamoto committed:
Update test_gneissweb_classification.py add gals

shahrokhDaijavad reviewed (approved):
LGTM
